### PR TITLE
Improve contour performance in WCSAxes by 10-1000x

### PR DESCRIPTION
**Pull request by astrofrog**
### Background

WCSAxes has a [mechanism](http://docs.astropy.org/en/stable/visualization/wcsaxes/overlays.html#contours) for overplotting contours from an image in a different coordinate system from the original image. This essentially boils down to:

```python
ax.contour(contour_data, transform=ax.get_transform(contour_wcs))
```

However, the way Matplotlib deals with this internally is to compute the contours on ``contour_data``, then to represent each line in the contour map as a Matplotlib ``Path`` (note - each line, not each level - there might be multiple lines per level), then to transform each path individually using the specified transform. In the case where coordinate conversions are needed, the transform includes calls to ``SkyCoord``, which we know is a little slow, especially if called once for each line in the contour map. For some contour maps I've made, there are sometimes thousands of individual lines.

### New approach

The approach implemented here is to overload ``contour`` and ``contourf``, and internally call them without transform, but to then modify the resulting ContourSet manually. We do this by extracting all the vertices of the lines, stacking them into a single array, transforming them, and splitting them again. This is far more efficient because we then only effectively call ``SkyCoord`` once.

### Performance

The following functions test the new approach, the old approach, and also test the performance without contours:

```python
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.axes import Axes

from astropy.io import fits
from astropy.wcs import WCS
from astropy.utils.data import get_pkg_data_filename

hdu_2mass = fits.open(get_pkg_data_filename('galactic_center/gc_2mass_h.fits'))[0]
hdu_msx = fits.open(get_pkg_data_filename('galactic_center/gc_msx_e.fits'))[0]


def new():
    ax = plt.subplot(1, 1, 1, projection=WCS(hdu_2mass.header))
    ax.imshow(hdu_2mass.data)
    ax.contour(hdu_msx.data, levels=np.logspace(-6, -3, 10),
               transform=ax.get_transform(WCS(hdu_msx.header)), colors='white')
    plt.savefig('new.png')


def old():
    ax = plt.subplot(1, 1, 1, projection=WCS(hdu_2mass.header))
    ax.imshow(hdu_2mass.data)
    Axes.contour(ax, hdu_msx.data, levels=np.logspace(-6, -3, 10),
                 transform=ax.get_transform(WCS(hdu_msx.header)), colors='white')
    plt.savefig('old.png')


def nocontour():
    ax = plt.subplot(1, 1, 1, projection=WCS(hdu_2mass.header))
    ax.imshow(hdu_2mass.data)
    plt.savefig('nocontour.png')
```

The results are as follows:

```python
In [10]: %timeit old()
14.5 s ± 710 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [11]: %timeit new()
198 ms ± 8.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [13]: %timeit nocontour()
125 ms ± 1.39 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

If we subtract the base time for the plot excluding the contour, the old approach took 14+ seconds, whereas now it takes ~70ms, a speedup of around 200x.

The exact speedup will depend a lot on the number of lines in the contour map. The worst case scenario (for the new implementation) would be if there was just a single contour line. If I take the example above and set ``levels=[5.8e-4]``, I get a single contour line. In that case, the timings look like:

```python
In [5]: %timeit old()
156 ms ± 3.65 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [4]: %timeit new()
145 ms ± 1.85 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [6]: %timeit nocontour()
153 ms ± 7.71 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Basically the differences are in the noise. So there should not really be any cases where this results in worse performance.

I've added a couple of benchmark to astropy-benchmarks: https://github.com/astropy/astropy-benchmarks/pull/58

### CODE REVIEW

File: astropy/visualization/wcsaxes/utils.py [hunk 1]
```diff
@@ -134,3 +134,56 @@ def coord_type_from_ctype(ctype):
...
+    # Transform all coordinates in one go
+    vertices = transform.transform(vertices)
+
+    # Split up into levels again
+    vertices = np.split(vertices, pos_level)
+
+    # Now re-populate the segments in the line collections
+    for ilevel, vert in enumerate(vertices):
+        vert = np.split(vert, pos_segments[ilevel])
+        for iseg in range(len(vert)):
```
Comment (larrybradley) [inline astropy/visualization/wcsaxes/utils.py]: I *think* using `enumerate` here would be *slightly* faster, e.g. something like
```
for iseg, ivert in vert:
    all_paths[ilevel][iseg].vertices = ivert
```

**Comment by larrybradley** [comment]
@astrofrog When will the new contour benchmarks be run?  I don't see them yet.  I assume this PR shouldn't be merged until the benchmarks have been run at least once with the old code?

File: astropy/visualization/wcsaxes/core.py [hunk 2]
```diff
@@ -196,6 +196,46 @@ def imshow(self, X, *args, **kwargs):
 
         return super().imshow(X, *args, **kwargs)
 
+    def contour(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (Cadair) [reply astropy/visualization/wcsaxes/core.py]: Would it be worth sticking `functools.wraps` on this so you have the function signature and stuff?

[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/visualization/wcsaxes/core.py]: Good point - I've written some custom docstrings that explicitly refer to contour and contourf via intersphinx for information about valid arguments.

File: astropy/visualization/wcsaxes/core.py [hunk 3]
```diff
@@ -196,6 +196,46 @@ def imshow(self, X, *args, **kwargs):
...
+
+        # In Matplotlib, when calling contour() with a transform, each
+        # individual path in the contour map is transformed separately. However,
+        # this is much too slow for us since each call to the transforms results
+        # in an Astropy coordinate transformation, which has a non-negligeable
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/visualization/wcsaxes/core.py]: Fixed

File: astropy/visualization/wcsaxes/utils.py [hunk 4]
```diff
@@ -134,3 +134,56 @@ def coord_type_from_ctype(ctype):
...
+
+
+def transform_contour_set_inplace(cset, transform):
+    """
+    Transform a contour set in-place using a specified
+    :class:`matplotlib.transform.Transform`
+
+    Using transforms with the native Matplotlib contour/contourf can be slow if
+    the transforms have a non-negligible overhead (which is the case for
+    WCS/Skycoord transforms) since the transform is called for each individual
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/visualization/wcsaxes/utils.py]: Fixed

File: astropy/visualization/wcsaxes/utils.py [hunk 5]
```diff
@@ -134,3 +134,56 @@ def coord_type_from_ctype(ctype):
...
+    # all paths and transform them in one go. The pos_level list helps us keep
+    # track of where the set of segments for each overall contour level ends.
+    # The pos_segments list helps us keep track of where each segmnt ends for
+    # each contour level.
+    all_paths = []
+    pos_level = []
+    pos_segments = []
+
+    for collection in cset.collections:
+
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/visualization/wcsaxes/utils.py]: Removed

File: astropy/visualization/wcsaxes/utils.py (hunk 1 above)
Comment (astrofrog) [reply astropy/visualization/wcsaxes/utils.py]: Fixed

**Comment by astrofrog** [comment]
@larrybradley - thanks for the review! I'll check today why the benchmarks haven't run yet.

File: astropy/visualization/wcsaxes/core.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/visualization/wcsaxes/core.py]: @Cadair - Actually if we didn't set any docstring I think it just inherits it? But I'd rather go with the approach done here now so we don't get random doc failures when Matplotlib change their docstring.

[the remark this answers is not in the document]
Comment (Cadair) [reply astropy/visualization/wcsaxes/core.py]: @astrofrog I wasn't actually thinking for the docstring, I was thinking for the function signature.

[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy/visualization/wcsaxes/core.py]: @Cadair ah, that won't be very interesting though:

```
Signature: ax.contour(*args, data=None, **kwargs)
```

(that's the matplotlib one!)

[the remark this answers is not in the document]
Comment (larrybradley) [reply astropy/visualization/wcsaxes/core.py]: 😄

**Comment by larrybradley** [comment]
@astrofrog Let me know when the benchmarks have been run and I'll merge this.

**Comment by astrofrog** [comment]
@larrybradley - they have now run:

http://www.astropy.org/astropy-benchmarks/#visualization.wcsaxes.time_contour_with_transform

Thanks!

File: astropy/visualization/wcsaxes/core.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Cadair) [reply astropy/visualization/wcsaxes/core.py]: :woman_facepalming:
